File: android-mindsherpa/app/src/main/java/com/skillvergence/mindsherpa/data/api/TeacherApiService.kt
package com.skillvergence.mindsherpa.data.api

import retrofit2.Response
import retrofit2.http.*

/**
 * Teacher API Service for Android
 * Corresponds to TeacherAPIService.swift in iOS
 */
interface TeacherApiService {

    @POST("teacher/validate-code")
    suspend fun validateTeacherCode(
        @Body request: TeacherCodeRequest
    ): Response<TeacherValidationResponse>

    @GET("teacher/school/{schoolId}/config")
    suspend fun getSchoolConfig(
        @Path("schoolId") schoolId: String
    ): Response<SchoolConfigResponse>

    @GET("teacher/school/{schoolId}/students")
    suspend fun getStudentRoster(
        @Path("schoolId") schoolId: String,
        @Query("level") level: String? = null,
        @Query("sortBy") sortBy: String = "name",
        @Query("order") order: String = "asc"
    ): Response<StudentRosterResponse>

    @GET("teacher/students/{studentId}/progress")
    suspend fun getStudentProgress(
        @Path("studentId") studentId: String
    ): Response<StudentProgressResponse>

    @GET("teacher/school/{schoolId}/certificates")
    suspend fun getCertificates(
        @Path("schoolId") schoolId: String,
        @Query("status") status: String = "all"
    ): Response<CertificatesResponse>

    @POST("teacher/certificates/{certId}/approve")
    suspend fun approveCertificate(
        @Path("certId") certId: String,
        @Body request: CertificateActionRequest
    ): Response<CertificateActionResponse>

    @GET("teacher/school/{schoolId}/code-usage")
    suspend fun getCodeUsage(
        @Path("schoolId") schoolId: String
    ): Response<CodeUsageResponse>

    @POST("teacher/school/{schoolId}/generate-codes")
    suspend fun generateCodes(
        @Path("schoolId") schoolId: String,
        @Body request: CodeGenerationRequest
    ): Response<CodeGenerationResponse>

    @POST("teacher/generate-class-code")
    suspend fun generateClassCode(
        @Body request: ClassCodeGenerationRequest
    ): Response<ClassCodeResponse>

    @GET("teacher/class-code")
    suspend fun getCurrentClassCode(
        @Query("teacherId") teacherId: String
    ): Response<ClassCodeResponse>

    @POST("student/join-class")
    suspend fun joinClass(
        @Body request: JoinClassRequest
    ): Response<JoinClassResponse>
}

// MARK: - Request Models

data class TeacherCodeRequest(
    val code: String,
    val schoolId: String
)

data class CertificateActionRequest(
    val action: String,
    val teacherId: String
)

data class CodeGenerationRequest(
    val type: String,
    val count: Int
)

data class ClassCodeGenerationRequest(
    val teacherId: String,
    val teacherCode: String,
    val regenerate: Boolean = false
)

data class JoinClassRequest(
    val classCode: String,
    val studentName: String,
    val studentEmail: String
)

// MARK: - Response Models

data class TeacherValidationResponse(
    val success: Boolean,
    val teacher: TeacherDetails?,
    val error: String?
)

data class TeacherDetails(
    val id: String,
    val name: String,
    val email: String,
    val school: String,
    val schoolId: String,
    val department: String,
    val program: String,
    val classCode: String
)

data class SchoolInfo(
    val id: String,
    val name: String,
    val district: String,
    val program: String,
    val instructor: InstructorInfo
)

data class InstructorInfo(
    val id: String,
    val name: String,
    val email: String,
    val department: String?
)

data class SchoolConfigResponse(
    val school: SchoolConfig
)

data class SchoolConfig(
    val id: String,
    val name: String,
    val district: String,
    val program: String,
    val instructor: InstructorInfo,
    val xpThreshold: Int,
    val bulkLicenses: Int,
    val districtLicenses: Int
)

data class StudentRosterResponse(
    val students: List<ApiStudent>,
    val summary: StudentSummary
)

data class ApiStudent(
    val id: String,
    val name: String,
    val email: String,
    val courseLevel: String,
    val totalXP: Int,
    val currentLevel: Int,
    val completedCourses: Int,
    val lastActive: String,
    val streak: Int
)

data class StudentSummary(
    val totalStudents: Int,
    val activeToday: Int,
    val avgXP: Int,
    val totalCompletedCourses: Int,
    val avgCompletionRate: Int
)

data class StudentProgressResponse(
    val student: ApiStudent,
    val devices: List<ApiDevice>,
    val courseProgress: List<CourseProgress>,
    val weeklyActivity: List<WeeklyActivity>,
    val achievements: List<Achievement>
)

data class CourseProgress(
    val courseId: String,
    val courseName: String,
    val progress: Int,
    val completedVideos: Int,
    val totalVideos: Int,
    val timeSpent: Int
)

data class WeeklyActivity(
    val week: String,
    val xpEarned: Int,
    val timeSpent: Int
)

data class Achievement(
    val title: String,
    val earnedDate: String
)

data class ApiDevice(
    val deviceId: String,
    val deviceName: String,
    val platform: String,
    val appVersion: String,
    val lastSeen: String,
    val isActive: Boolean
)

data class CertificatesResponse(
    val certificates: List<ApiCertificate>,
    val summary: CertificateSummary
)

data class ApiCertificate(
    val id: String,
    val studentId: String,
    val studentName: String,
    val courseTitle: String,
    val completedDate: String,
    val status: String,
    val approvedBy: String?,
    val approvedDate: String?
)

data class CertificateSummary(
    val total: Int,
    val pending: Int,
    val approved: Int
)

data class CertificateActionResponse(
    val success: Boolean,
    val certificate: ApiCertificate
)

data class CodeUsageResponse(
    val usage: CodeUsage,
    val generatedCodes: GeneratedCodes
)

data class CodeUsage(
    val basicCodes: Int,
    val premiumCodes: Int,
    val friendCodes: Int
)

data class GeneratedCodes(
    val basic: List<String>,
    val premium: List<String>
)

data class CodeGenerationResponse(
    val success: Boolean,
    val codes: List<String>,
    val type: String,
    val count: Int,
    val generatedAt: String
)

data class ClassCodeResponse(
    val success: Boolean,
    val classCode: String?,
    val teacherInfo: TeacherInfo?,
    val error: String?
)

data class TeacherInfo(
    val id: String,
    val name: String,
    val email: String,
    val schoolName: String,
    val programName: String
)

data class JoinClassResponse(
    val success: Boolean,
    val teacherInfo: TeacherInfo?,
    val accessLevel: String?,
    val error: String?
)

File: ios/mindsherpa/mindsherpa/TeacherAPIService.swift
//
//  TeacherAPIService.swift
//  mindsherpa
//
//  Created by Claude on 9/16/25.
//

import Foundation
import Combine

@MainActor
class TeacherAPIService: ObservableObject {
    static let shared = TeacherAPIService()

    private let baseURL: String
    private var cancellables = Set<AnyCancellable>()

    private init() {
        // Use AppConfig for consistent URL management
        self.baseURL = AppConfig.apiURL
    }

    // MARK: - Teacher Authentication

    func validateTeacherCode(_ code: String, schoolId: String = "") async throws -> TeacherValidationResponse {
        let url = URL(string: "\(baseURL)/teacher/validate-code")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = TeacherCodeRequest(code: code, schoolId: schoolId)
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw TeacherAPIError.invalidResponse
        }

        if httpResponse.statusCode == 401 {
            throw TeacherAPIError.invalidCode
        } else if httpResponse.statusCode != 200 {
            throw TeacherAPIError.serverError("HTTP \(httpResponse.statusCode)")
        }

        return try JSONDecoder().decode(TeacherValidationResponse.self, from: data)
    }

    // MARK: - School Configuration

    func getSchoolConfig(schoolId: String) async throws -> SchoolConfigResponse {
        let url = URL(string: "\(baseURL)/teacher/school/\(schoolId)/config")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode(SchoolConfigResponse.self, from: data)
    }

    // MARK: - Student Management

    func getStudentRoster(schoolId: String, level: String? = nil, sortBy: String = "name", order: String = "asc") async throws -> StudentRosterResponse {
        var components = URLComponents(string: "\(baseURL)/teacher/school/\(schoolId)/students")!
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "sortBy", value: sortBy),
            URLQueryItem(name: "order", value: order)
        ]

        if let level = level {
            queryItems.append(URLQueryItem(name: "level", value: level))
        }

        components.queryItems = queryItems

        let (data, _) = try await URLSession.shared.data(from: components.url!)
        return try JSONDecoder().decode(StudentRosterResponse.self, from: data)
    }

    func getStudentProgress(studentId: String) async throws -> StudentProgressResponse {
        let url = URL(string: "\(baseURL)/teacher/students/\(studentId)/progress")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode(StudentProgressResponse.self, from: data)
    }

    // MARK: - Certificate Management

    func getCertificates(schoolId: String, status: String = "all") async throws -> CertificatesResponse {
        let url = URL(string: "\(baseURL)/teacher/school/\(schoolId)/certificates?status=\(status)")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode(CertificatesResponse.self, from: data)
    }

    func approveCertificate(certId: String, action: String, teacherId: String) async throws -> CertificateActionResponse {
        let url = URL(string: "\(baseURL)/teacher/certificates/\(certId)/approve")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = CertificateActionRequest(action: action, teacherId: teacherId)
        request.httpBody = try JSONEncoder().encode(body)

        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode(CertificateActionResponse.self, from: data)
    }

    // MARK: - Code Management

    func getCodeUsage(schoolId: String) async throws -> CodeUsageResponse {
        let url = URL(string: "\(baseURL)/teacher/school/\(schoolId)/code-usage")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode(CodeUsageResponse.self, from: data)
    }

    func generateCodes(schoolId: String, type: String, count: Int) async throws -> CodeGenerationResponse {
        let url = URL(string: "\(baseURL)/teacher/school/\(schoolId)/generate-codes")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = CodeGenerationRequest(type: type, count: count)
        request.httpBody = try JSONEncoder().encode(body)

        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode(CodeGenerationResponse.self, from: data)
    }
}

// MARK: - Data Models

struct TeacherCodeRequest: Codable {
    let code: String
    let schoolId: String
}

struct TeacherValidationResponse: Codable {
    let success: Bool
    let teacher: TeacherInfo?
    let error: String?

    // Computed property for backward compatibility
    var valid: Bool { success }
}

struct TeacherInfo: Codable {
    let id: String
    let name: String
    let email: String
    let school: String
    let schoolId: String
    let department: String?
    let program: String?
    let classCode: String?
}

struct SchoolInfo: Codable {
    let id: String
    let name: String
    let district: String
    let program: String
    let instructor: InstructorInfo
}

struct InstructorInfo: Codable {
    let id: String
    let name: String
    let email: String
    let department: String?
}

struct SchoolConfigResponse: Codable {
    let school: SchoolConfig
}

struct SchoolConfig: Codable {
    let id: String
    let name: String
    let district: String
    let program: String
    let instructor: InstructorInfo
    let xpThreshold: Int
    let bulkLicenses: Int
    let districtLicenses: Int
}

struct StudentRosterResponse: Codable {
    let students: [APIStudent]
    let summary: StudentSummary
}

struct APIStudent: Codable, Identifiable {
    let id: String
    let name: String
    let email: String
    let courseLevel: String
    let totalXP: Int
    let currentLevel: Int
    let completedCourses: Int
    let lastActive: String
    let streak: Int
}

struct StudentSummary: Codable {
    let totalStudents: Int
    let activeToday: Int
    let avgXP: Int
    let totalCompletedCourses: Int
    let avgCompletionRate: Int
}

struct StudentProgressResponse: Codable {
    let student: APIStudent
    let courseProgress: [CourseProgress]
    let weeklyActivity: [WeeklyActivity]
    let achievements: [Achievement]
}

struct CourseProgress: Codable {
    let courseId: String
    let courseName: String
    let progress: Int
    let completedVideos: Int
    let totalVideos: Int
    let timeSpent: Int
}

struct WeeklyActivity: Codable {
    let week: String
    let xpEarned: Int
    let timeSpent: Int
}

struct Achievement: Codable {
    let title: String
    let earnedDate: String
}

struct CertificatesResponse: Codable {
    let certificates: [APICertificate]
    let summary: CertificateSummary
}

struct APICertificate: Codable, Identifiable {
    let id: String
    let studentId: String
    let studentName: String
    let courseTitle: String
    let completedDate: String
    let status: String
    let approvedBy: String?
    let approvedDate: String?
}

struct CertificateSummary: Codable {
    let total: Int
    let pending: Int
    let approved: Int
}

struct CertificateActionRequest: Codable {
    let action: String
    let teacherId: String
}

struct CertificateActionResponse: Codable {
    let success: Bool
    let certificate: APICertificate
}

struct CodeUsageResponse: Codable {
    let usage: CodeUsage
    let generatedCodes: GeneratedCodes
}

struct CodeUsage: Codable {
    let basicCodes: Int
    let premiumCodes: Int
    let friendCodes: Int
}

struct GeneratedCodes: Codable {
    let basic: [String]
    let premium: [String]
}

struct CodeGenerationRequest: Codable {
    let type: String
    let count: Int
}

struct CodeGenerationResponse: Codable {
    let success: Bool
    let codes: [String]
    let type: String
    let count: Int
    let generatedAt: String
}

// MARK: - Error Handling

enum TeacherAPIError: Error, LocalizedError {
    case invalidCode
    case invalidResponse
    case serverError(String)
    case networkError

    var errorDescription: String? {
        switch self {
        case .invalidCode:
            return "Invalid teacher access code"
        case .invalidResponse:
            return "Invalid server response"
        case .serverError(let message):
            return "Server error: \(message)"
        case .networkError:
            return "Network connection error"
        }
    }
}